### ISSUE TITLE
Next 9  - Using functional components as child of <Link/> causes ref-warnings

### ISSUE THREAD

**Issue by nickluger** [zeit/next.js#7915]
# Bug report

## Describe the bug

Using a functional component as a child of `<Link/>` causes:
```
Warning: Function components cannot be given refs. Attempts to access this ref will fail. Did you mean to use React.forwardRef()?
```
This happened only after upgrading from v8 => v9.

## To Reproduce

Reproduction repo: https://github.com/nickluger/nextjs-link-ref-functional-comp-warning

## Expected behavior

`<Link/>` should work with functional components, too / not show a warning.

## System information

- Version of Next.js: 9.0.1

**Comment by mustaphaturhan** [comment]
Same here. I am having error If I use custom component like this:

```html
        <Link href="/href">
          <Button
            type="custom"
            color="#FCFCFC"
          >
             buttonText
          </Button>
        </Link>
```
This is only happening when you use a component as a child. These don't give a warning for example:
```html
        <Link href="/href">
          <button
            type="custom"
            color="#FCFCFC"
          >
            buttonText
          </button>
        </Link>

        <Link href="/href">
          <a>
            <Button
              type="custom"
              color="#FCFCFC"
            >
              buttonText
            </Button>
          </a>
        </Link>
```

**Comment by joaogarin** [comment]
Is there a way around this bug? we are blocked from going to 9 for this one when some features we would really like to leverage like the typescript support. Just noticed it bumped to 9.0.3 :/

**Comment by nickluger** [comment]
You could try muting it temporarily , like: 
```js
// TODO: Muting error, fix as soon as zeit/next.js/issues/7915 resolved
const originalError = console.error;

console.error = (...args) => {
  if (/Warning.*Function components cannot be given refs/.test(args[0])) {
    return;
  }
  originalError.call(console, ...args);
};
```

**Comment by focux** [comment]
I'm having the same issue here.

**Comment by paulboony** [comment]
If you use `<Link/>` to wrap your custom component, you could forward `ref` like this:

```jsx
const CustomComponent = React.forwardRef(function CustomComponent(props, ref) {
  return (
    <div/>
  );
});
```

```jsx
<Link href={"/"}>
  <CustomComponent/>
</Link>
```

The warning goes away.

**Comment by th0th** [comment]
I have updated to `9.0.4-canary.2` but I still get the warning on the console. Am I missing something?

```
Warning: Function components cannot be given refs. Attempts to access this ref will fail. Did you mean to use React.forwardRef()?

Check the render method of `Link`.
```

**Comment by mustaphaturhan** [comment]
@th0th You should wrap your component in `React.forwardRef`. Like this one:

```js
import React from 'react'
import Link from 'next/link'

const CustomComponent = React.forwardRef((props, ref) => (
  <a ref={ref} {...props}>
    Click
  </a>
))

export default () => (
  <Link href='/' passHref>
    <CustomComponent/>
  </Link>
)
```

**Comment by nickluger** [comment]
Wrap all functional components used within a `Link` **manually**?   Shouldn't there be a easier, less repetitive solution for this? At least something like `<Link href="/" forwardRef>...</Link>`?

**Comment by tm1000** [comment]
Agree with @nickluger this seems counter-intuitive.

**Comment by srghma** [comment]
Should all other libraries implement `forwardRef` attribute in the same manner? Or maybe react creators thought that developers will use `React.forwardRef` function instead of implementing `forwardRef` attribute?

**Comment by DonovanCharpin** [comment]
Yes agree with @nickluger, it's a pain to do this manually for all the components components inside a `Link`.

**Comment by nickluger** [comment]
Should we reopen and rename this issue (@ijjk) or did someone already open a new one?

**Comment by bschwartz757** [comment]
> Same here. I am having error If I use custom component like this:
> 
> ```
>         <Link href="/href">
>           <Button
>             type="custom"
>             color="#FCFCFC"
>           >
>              buttonText
>           </Button>
>         </Link>
> ```
> 
> This is only happening when you use a component as a child. These don't give a warning for example:
> 
> ```
>         <Link href="/href">
>           <button
>             type="custom"
>             color="#FCFCFC"
>           >
>             buttonText
>           </button>
>         </Link>
> 
>         <Link href="/href">
>           <a>
>             <Button
>               type="custom"
>               color="#FCFCFC"
>             >
>               buttonText
>             </Button>
>           </a>
>         </Link>
> ```

This will most likely throw a console error:

`Warning: validateDOMNesting(...): <a> cannot appear as a descendant of <a>.`

**Comment by jon64digital** [comment]
I just hit this issue when upgrading Next JS. Is there a fix on the way?

I'm not sure I understand how to use forwardRef and wrap my component. My button.js file currently exports the following....

`
const Button = (props) => {
    return (<stuff />)
}

export default React.memo(Button)


`

**Comment by nickluger** [comment]
See above, I opened a follow-up here: https://github.com/zeit/next.js/issues/8962

**Comment by bschwartz757** [comment]
> I just hit this issue when upgrading Next JS. Is there a fix on the way?
> 
> I'm not sure I understand how to use forwardRef and wrap my component. My button.js file currently exports the following....
> 
> ```
> const Button = (props) => {
>     return (<stuff />)
> }
> 
> export default React.memo(Button)
> ```

In this case, you should be able to just do
```
import {forwardRef} from 'react';
const Button = forwardRef(props, ref) => {
    return (<stuff ref={ref} />)
}
```
for the forwardRef part. The docs also mention https://nextjs.org/docs#forcing-the-link-to-expose-href-to-its-child, though I haven't tried that. I've found that doing this

```
            <Link
              href={href}
            >
              <a>
                {text}
              </a>
            </Link>
```
works (despite looking very odd.)

**Comment by Vadorequest** [comment]
Encountering this issue in 9.0.2

Using `<a></a>` around may work, but will trigger other error like "must have href", etc. when using a11y and similar.

**Comment by hill84** [comment]
What about this solution using a react **fragment** as a bearing wrapper?
```
<Link href={href}>
  <Fragment>
      <MyFunctionalComponent ref={myRef} />
  </Fragment>
</Link>
```
Where `<MyFunctionalComponent>` forwards the ref:
```
const MyFunctionalComponent = forwardRef(props, ref) => (
  <div ref={ref} />
);
```
This doesn't add any additional HTML wrapper in the output: `<div />`

**Comment by Vadorequest** [comment]
@hill84 Interesting, but where does `myRef` comes from in your example?

**Comment by hill84** [comment]
@Vadorequest I've added more details to my previous comment.

**Comment by tim-codes** [comment]
Following @hill84's suggestion was a nice workaround:
```const ButtonLink: FC<ButtonLinkProps> = ({ children, href, ...props }) => (
  <Link href={href}>
    <><Button link>{children}</Button></>
  </Link>
);```

**Comment by namlq93** [comment]
> Following @hill84's suggestion was a nice workaround:
> 
> ```
> const ButtonLink: FC<ButtonLinkProps> = ({ children, href, ...props }) => (
>   <Link href={href}>
>     <><Button link {...props}>{children}</Button></>
>   </Link>
> );
> ```

I use `<><Component /></>` but Link can not action. I have to use `<div></div>` instead of.

**Comment by andy-pro** [comment]
I suggest a solution where a function returns an object with props  instead of a component :

```js
const propsUserBtn = user => ({
    className: "h-10 w-10 font-black rounded-full bg-orange-200 text-teal-700",
    title: user.first_name + ' ' + user.last_name,
    children: (user.first_name.charAt(0) + user.last_name.charAt(0)).toUpperCase()
})
```

```js
<Link href='/profile'>
    <button { ...propsUserBtn(user) } />   
</Link>
```

Checked, it works!

**Comment by mikeyrt16** [comment]
Why is this ref even needed? Is it to guarantee the correct placement of the `href` in the underlying wrapped component when `passHref` is used in the `<Link />`?

This is super annoying as my custom <Button /> component has the `<a>` at the root and I can see it's getting href applied to it correctly anyway... I'd much rather have it enforced that you need your underlying custom component to have the anchor at the root and avoid this whole `ref` design desicion.

`React.forwardRef` is a horribly cumbersome thing to force people to implement and is playing hell with my linters 😞

**Comment by mikeyrt16** [comment]
Or at least have it be an opt in feature if you need to specify the placement of the `href` in your component. `useRef`?

**Comment by LauraBeatris** [comment]
I don't think that having to create another module to forward the ref to a custom component is a good solution and also it's too repetitive. 

The way that I'm using to handle this issue is by wrapping the custom component with a ``div`` element like suggested by @namlq93 but it also doesn't seem to be clean...

**Comment by hothvictor** [comment]
> Same here. I am having error If I use custom component like this:
> 
> ```
>         <Link href="/href">
>           <Button
>             type="custom"
>             color="#FCFCFC"
>           >
>              buttonText
>           </Button>
>         </Link>
> ```
> 
> This is only happening when you use a component as a child. These don't give a warning for example:
> 
> ```
>         <Link href="/href">
>           <button
>             type="custom"
>             color="#FCFCFC"
>           >
>             buttonText
>           </button>
>         </Link>
> 
>         <Link href="/href">
>           <a>
>             <Button
>               type="custom"
>               color="#FCFCFC"
>             >
>               buttonText
>             </Button>
>           </a>
>         </Link>
> ```

It works, thank mate!!

**Comment by TomasSestak** [comment]
still there

**Comment by cgarrovillo** [comment]
https://nextjs.org/docs/api-reference/next/link#if-the-child-is-a-function-component

**Comment by cristian-eriomenco** [comment]
This is sad.

**Comment by jakehewitt** [comment]
I was watching the NextJS conference the other day thinking "Wow Next is the future". I really like Next and the Vercel ecosystem and I want to honestly be able to say that and be an advocate for Next. However, issues like these being closed and not addressed make it hard to do so. I've seen this type of pattern multiple times throughout the issues here, specifically with the Router.

There's 3 solutions to this problem from the user side and none of them are great DX. 

1. The recommended way React.forwardRef. wrap the component in `React.forwardRef`. This can be very repetitive
2. Wrap the functional component in a tag like `div`, `span`, `a`, etc. 
3. Instead of Link, pull in useRouter and use `router.push`

No matter which solution you use, they're all strange and require comments linking to this issue or the [docs](https://nextjs.org/docs/api-reference/next/link#if-the-child-is-a-function-component) to explain to coworkers/contributors why its needed to prevent accidental removal/cleanup. 

I think this issue should be reopened for discussion.

**Comment by progammer-rkt** [comment]
I recently came across this issue and the way i solved is like this:

```
<Link href="/">
    <>
        <CustomComponent />
    </>
</Link>
```

So React Fragment seems to work in this case.

**Comment by huntedman** [comment]
> I recently came across this issue and the way i solved is like this:
> 
> ```
> <Link href="/">
>     <>
>         <CustomComponent />
>     </>
> </Link>
> ```
> 
> So React Fragment seems to work in this case.

Works for me aswell. Wondering if it has any implications...

**Comment by maksimf** [comment]
Warning does go away when using react fragment, but link doesn't work for me

**Comment by Fjollsfinn** [comment]
Next version: 10.0.1. The problem still persists.

**Comment by bragle** [comment]
@progammer-rkt  @huntedman @maksimf Just keep in mind that [`passHref`](https://nextjs.org/docs/api-reference/next/link#if-the-child-is-a-custom-component-that-wraps-an-a-tag) wont work with this solution...

**Comment by bayraak** [comment]
I'm just creating a wrapper component for generic usage:

```
const CustomLinkWrapper = React.forwardRef((props: any, ref) => (
    <a ref={ref} {...props}>
        {props.children}
    </a>
));
```

and then I'm using this inside the Link like the following:

```
<Link
          href={`/[slug]?slug=${item.slug}`}
          as={`/${item.slug}`}
           >
                  <CustomLinkWrapper>
                    <ProductCard {...item}></ProductCard>
                  </CustomLinkWrapper>
</Link>
```

**Comment by devthejo** [comment]
```js
import NextLink from "next/link";

function LinkWrap({ children, refAs }, ref) {
  const childProps = {};
  if (refAs) {
    childProps[refAs] = ref;
  }
  return (
    <>
      {React.isValidElement(children)
        ? React.cloneElement(children, childProps)
        : null}
    </>
  );
}

const LinkWrapper = React.forwardRef(LinkWrap);

function Link({ refAs, children, ...props }) {
  return (
    <NextLink {...props}>
      <LinkWrapper refAs={refAs}>{children}</LinkWrapper>
    </NextLink>
  );
}
```

and use created `Link` component instead of next.js one

**Comment by devthejo** [comment]
```js
import NextLink from "next/link";

function LinkWrap({ children, refAs, ...props }, ref) {
  if (refAs) {
    props[refAs] = ref;
  }
  return (
    <>
      {React.isValidElement(children)
        ? React.cloneElement(children, props)
        : null}
    </>
  );
}

const LinkWrapper = React.forwardRef(LinkWrap);

function Link({ refAs, children, ...props }) {
  return (
    <NextLink {...props}>
      <LinkWrapper refAs={refAs}>{children}</LinkWrapper>
    </NextLink>
  );
}
```

and use created `Link` component instead of next.js one

**Comment by machineghost** [comment]
```
<Link href="/">
    <>
        <CustomComponent />
    </>
</Link>
```

This is simultaneously genius AND incredibly ... I mean really, really ... stupid.

If I were on the Next.js team I would  be _embarassed_ that I was forcing people to resort to this.

**Comment by Otoris** [comment]
@machineghost Be nice - there are human people on the receiving end of your words.

**Comment by adamduncan** [comment]
If we've got a fair bit going on in the internals, and need to use our own functional component directly within Next's `Link`, is it bananas to create ourselves a proxy component that forwards on the refs and props?

```jsx
import { Button } from 'some-package';

const ButtonNext = React.forwardRef(({ children, ...rest }, ref) => {
  return (
    <span ref={ref}>
      <Button {...rest}>{children}</Button>
    </span>
  );
});


// ...

<Link href="/" passHref>
  <ButtonNext>Go home</ButtonNext>
</Link>
```

I.e. We can't attach a `ref` to our FC directly, so the wrapping span element takes it on for us.

**Comment by arthurvergacas** [comment]
Why is this issue closed? Was it solved? If so, where?

**Comment by EverStarck** [comment]
> Why is this issue closed? Was it solved? If so, where?
> I solved it with the <div></div> trick, but still seems hacky.

I have a next <Image .../> component an a <Link> </Link> and it throw the same warning. I put a div inside Link and it works but yes, it seems hacky

**Comment by sshanzel** [comment]
Didn't realize there are some issues like this lurking in NextJS. And by issue, I mean how the primary concern in this matter is seen as nothing to care about.

**Comment by MinyoungNa1997** [comment]
This is very much a riddle, do I spend time "fixing" this warning even though it's not a functional issue?

**Comment by nzlatinov** [comment]
After long and painful debugging I trial and error-ed the following solution for my case: 
Wrapping the functional component in a span automagically solves the issue for me. 

```
 <Link href={`?selectedStation=${id}`} scroll={false}>
  <span>
    {/* wrapping in a span avoids following bug: https://github.com/vercel/next.js/issues/7915  */}
    <CustomIconButton />
  </span>
</Link>
```

**Comment by OkkarMin** [comment]
The link that @cgarrovillo  provided fixed the issue for me. The warnings went away and my custom functional component is working well with the link.

https://nextjs.org/docs/api-reference/next/link#if-the-child-is-a-function-component

**Comment by MattCollyer** [comment]
Wrapping it in a <div> tag worked, but it didn't render the Links anchor tag which could mess up SEO. Instead, I made a custom Link wrapper which wraps the component in an anchor tag- as described here. https://jasonwatmore.com/post/2021/05/31/next-js-make-the-link-component-work-like-react-router-link

**Comment by chasoft** [comment]
I've just encountered this error today! Problem fixed with this guide [https://nextjs.org/docs/api-reference/next/link#if-the-child-is-a-function-component](https://nextjs.org/docs/api-reference/next/link#if-the-child-is-a-function-component)

**Comment by orangecoloured** [comment]
What's the use of forwarding the `ref`?
Can I simply have it like this?
`export const ComponentWithRef = React.forwardRef((props, ref) => <Component { ...props } />)`

**Comment by tri-bit** [comment]
If I wrap the functional component in <a></a> it works fine with no forwardRef... and I've been using this method for years - does this break proper internal routing in some way I'm not seeing? 

`<Link href="/internal-link"><a><MyFunctionalComponent/></a></Link>`

**Comment by theigwe** [comment]
So I followed this https://nextjs.org/docs/api-reference/next/link#if-the-child-is-a-custom-component-that-wraps-an-a-tag
And created this component for reuse when needed:

```
import Link from 'next/link';

const ButtonLink = ({ children, href, ...props }) => (
  <Link href={href} {...props} passHref>
    <a>{children}</a>
  </Link>
);

export { ButtonLink };
```

With this I have my component wrapped in `<a href=''>`

**Comment by manny-p** [comment]
> If I wrap the functional component in `<a></a>` it works fine with no forwardRef... and I've been using this method for years - does this break proper internal routing in some way I'm not seeing?
> 
> `<Link href="/internal-link"><a><MyFunctionalComponent/></a></Link>`

yep this worked for me too!

**Comment by jps** [comment]
This might help others, this is how I've solved with TS. It does feel like a big dirty hack having to add this `div` I don't need in there. Problem is I don't have access to the underlying `Card` component to expose the `ref` on it's interface similar to Adam above. 

```tsx

import Link from 'next/link';
import { forwardRef } from 'react';

export interface MediaCardProps extends CardProps {
  headline: string;
  children: string;
  href: string;
}

type MediaCardPropsWithOptionalHref = Omit<MediaCardProps, 'href'> & {
  href?: string;
};

const CardWrapped = forwardRef<any, MediaCardPropsWithOptionalHref>(
  ({ children, headline, href }, ref) => (
    <div ref={ref}>
      <Card layout="vertical" href={href}>
          {/* redacted */}
      </Card>
    </div>
  )
);

export const MediaCard = ({ href, children, headline }: MediaCardProps) => (
  <Link href={href} passHref>
    <CardWrapped headline={headline}>{children}</CardWrapped>
  </Link>
);


```

**Comment by ricardo-rp** [comment]
This is a small pain in the butt every time you need to use a <Link/>. Most components are functional components nowadays. 
Wrapping in <a> seems to be the most straightforward solution for most cases. I really think the nextjs team should think of a more convenient API though.

**Comment by balazsorban44** [comment]
This issue has been automatically locked due to no recent activity. If you are running into a similar issue, please create a new issue with the steps to reproduce. Thank you.
